Classes/EventSourcedRepository.php

```diff
@@ -85,32 +91,30 @@ public function findByIdentifier($identifier)
     public function save(AggregateRootInterface $aggregate)
     {
         try {
-            $stream = $this->eventStore->get($this->generateStreamName($aggregate->getAggregateIdentifier()));
+            $filter = EventStreamFilter::create()
+                ->setStreamName($this->getStreamName($aggregate->getAggregateIdentifier()));
+            $stream = $this->eventStore->get($filter);
         } catch (EventStreamNotFoundException $e) {
             $stream = new EventStream();
         }
 
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
         $stream->addEvents(...$uncommittedEvents);
 
-        $this->eventStore->commit($this->generateStreamName($aggregate->getAggregateIdentifier()), $stream, function ($version) use ($uncommittedEvents) {
-            /** @var EventTransport $eventTransport */
-            foreach ($uncommittedEvents as $eventTransport) {
-                // @todo metadata enrichment must be done in external service, with some middleware support
-                $eventTransport->getMetaData()->add(Metadata::VERSION, $version);
-                $this->eventBus->handle($eventTransport);
-            }
-        });
+        $this->eventHandlingService->publish($this->getStreamName($aggregate->getAggregateIdentifier()), $stream);
     }
 
     /**
      * @param string $identifier
      * @return string
-     * @todo find a more flexible way to generate stream name, need to be discussed
      */
-    protected function generateStreamName(string $identifier)
+    protected function getStreamName($identifier)
     {
-        $streamName = $this->aggregateClassName . '::' . $identifier;
-        return $streamName;
+        if ($this->streamNamePrefix === null) {
+            list($vendor, $package) = explode('\\', $this->aggregateClassName);
+            $aggregateName = substr($this->aggregateClassName, strrpos($this->aggregateClassName, '\\') + 1);
+            $this->streamNamePrefix = $vendor . ':' . $package . ':' . $aggregateName;
```

Comment:
That's dangerous. We have `PackageManagerInterface::getPackageByClassName()` to determine the package key from a class. Lets use this for the BC name